File: backend/javascript/src/services/models/document-chunk.ts
import { DocType, DocVersion, Permission } from './raw-document';

/**
 * =========================
 * 1. LLM 출력 전용 메타데이터
 * =========================
 * - OpenAI Structured Outputs(json_schema)와 1:1 대응
 * - "AI가 추론해서 낼 수 있는 값"만 포함
 */
export interface AiDocumentMetadata {
    title: string;             // 문서 제목
    docType: DocType;           // 문서 종류
    permission: Permission[];   // 접근 권한
    summary?: string;           // 문서 요약 (선택)
}

/**
 * =========================
 * 2. 최종 Chunk 메타데이터
 * =========================
 * - 파이프라인 전체를 거친 뒤
 * - 벡터 DB에 저장되는 "완성본 메타데이터"
 */
export interface ChunkMetadata extends AiDocumentMetadata {
    version: DocVersion;       // 'current' | 'old'
    updatedAt: string;         // 파일 수정일 (YYYY‑MM‑DD)
    sectionTitle: string;      // 현재 청크가 속한 섹션(헤더)
    filePath: string;          // 원본 파일 절대 경로
    chunkIndex: number;        // 청크 순번 (0,1,2,…)
    embedding?: number[];      // OpenAI 등에서 반환된 임베딩 배열
}


/**
 * =========================
 * 3. 실제 벡터 단위 문서
 * =========================
 * - 벡터 DB에 저장되는 최소 단위
 */
export interface DocumentChunk {
    id: string;                // 고유 ID : `${filePath}#${sectionTitle}#${chunkIndex}`
    text: string;              // 청크 본문
    metadata: ChunkMetadata;    // 위에서 정의한 메타데이터
}

File: backend/java/src/main/java/com/ragassistant/model/DocumentChunk.java
package com.ragassistant.model;

import java.util.ArrayList;
import java.util.List;

public class DocumentChunk {
    private String id;
    private String text;
    private ChunkMetadata metadata;

    public DocumentChunk() {
    }

    public DocumentChunk(String id, String text, ChunkMetadata metadata) {
        this.id = id;
        this.text = text;
        this.metadata = metadata;
    }

    public static DocumentChunkBuilder builder() {
        return new DocumentChunkBuilder();
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public ChunkMetadata getMetadata() {
        return metadata;
    }

    public void setMetadata(ChunkMetadata metadata) {
        this.metadata = metadata;
    }

    public static class ChunkMetadata {
        private String title;
        private String docType;
        private List<String> permission = new ArrayList<>();
        private String summary;
        private String version;
        private String updatedAt;
        private String sectionTitle;
        private String filePath;
        private int chunkIndex;
        private List<Double> embedding = new ArrayList<>();

        public ChunkMetadata() {
        }

        public static ChunkMetadataBuilder builder() {
            return new ChunkMetadataBuilder();
        }

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public String getDocType() {
            return docType;
        }

        public void setDocType(String docType) {
            this.docType = docType;
        }

        public List<String> getPermission() {
            return permission;
        }

        public void setPermission(List<String> permission) {
            this.permission = permission;
        }

        public String getSummary() {
            return summary;
        }

        public void setSummary(String summary) {
            this.summary = summary;
        }

        public String getVersion() {
            return version;
        }

        public void setVersion(String version) {
            this.version = version;
        }

        public String getUpdatedAt() {
            return updatedAt;
        }

        public void setUpdatedAt(String updatedAt) {
            this.updatedAt = updatedAt;
        }

        public String getSectionTitle() {
            return sectionTitle;
        }

        public void setSectionTitle(String sectionTitle) {
            this.sectionTitle = sectionTitle;
        }

        public String getFilePath() {
            return filePath;
        }

        public void setFilePath(String filePath) {
            this.filePath = filePath;
        }

        public int getChunkIndex() {
            return chunkIndex;
        }

        public void setChunkIndex(int chunkIndex) {
            this.chunkIndex = chunkIndex;
        }

        public List<Double> getEmbedding() {
            return embedding;
        }

        public void setEmbedding(List<Double> embedding) {
            this.embedding = embedding;
        }
    }

    public static class DocumentChunkBuilder {
        private String id;
        private String text;
        private ChunkMetadata metadata;

        public DocumentChunkBuilder id(String id) {
            this.id = id;
            return this;
        }

        public DocumentChunkBuilder text(String text) {
            this.text = text;
            return this;
        }

        public DocumentChunkBuilder metadata(ChunkMetadata metadata) {
            this.metadata = metadata;
            return this;
        }

        public DocumentChunk build() {
            return new DocumentChunk(id, text, metadata);
        }
    }

    public static class ChunkMetadataBuilder {
        private String title;
        private String docType;
        private List<String> permission;
        private String summary;
        private String version;
        private String updatedAt;
        private String sectionTitle;
        private String filePath;
        private int chunkIndex;
        private List<Double> embedding;

        public ChunkMetadataBuilder title(String title) {
            this.title = title;
            return this;
        }

        public ChunkMetadataBuilder docType(String docType) {
            this.docType = docType;
            return this;
        }

        public ChunkMetadataBuilder permission(List<String> permission) {
            this.permission = permission;
            return this;
        }

        public ChunkMetadataBuilder summary(String summary) {
            this.summary = summary;
            return this;
        }

        public ChunkMetadataBuilder version(String version) {
            this.version = version;
            return this;
        }

        public ChunkMetadataBuilder updatedAt(String updatedAt) {
            this.updatedAt = updatedAt;
            return this;
        }

        public ChunkMetadataBuilder sectionTitle(String sectionTitle) {
            this.sectionTitle = sectionTitle;
            return this;
        }

        public ChunkMetadataBuilder filePath(String filePath) {
            this.filePath = filePath;
            return this;
        }

        public ChunkMetadataBuilder chunkIndex(int chunkIndex) {
            this.chunkIndex = chunkIndex;
            return this;
        }

        public ChunkMetadataBuilder embedding(List<Double> embedding) {
            this.embedding = embedding;
            return this;
        }

        public ChunkMetadata build() {
            ChunkMetadata m = new ChunkMetadata();
            m.setTitle(title);
            m.setDocType(docType);
            m.setPermission(permission);
            m.setSummary(summary);
            m.setVersion(version);
            m.setUpdatedAt(updatedAt);
            m.setSectionTitle(sectionTitle);
            m.setFilePath(filePath);
            m.setChunkIndex(chunkIndex);
            m.setEmbedding(embedding);
            return m;
        }
    }
}
